pkg/util/bytesbufpool/bytesbufpool.go

```diff
@@ -0,0 +1,58 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bytesbufpool
+
+import (
+	"bytes"
+	"sync"
+)
+
+var pool = sync.Pool{}
+
+const defaultCap = 100
+
+// Get a buffer with default capacity.
+func Get() *bytes.Buffer {
+	return GetWithCap(defaultCap)
+}
+
+// GetWithCap get a buffer with at least n bytes can be written to it without another allocation.
+// If n is negative, GetWithCap will panic.
+func GetWithCap(n int) *bytes.Buffer {
+	val := pool.Get()
+	if val != nil {
+		buf := val.(*bytes.Buffer)
+		if buf.Cap() < n {
+			buf.Grow(n)
+		}
+		return buf
+	}
+	return bytes.NewBuffer(make([]byte, 0, n))
+}
+
+func Put(buf *bytes.Buffer) {
+	// Proper usage of a sync.Pool requires each entry to have approximately
+	// the same memory cost. To obtain this property when the stored type
+	// contains a variably-sized buffer, we add a hard limit on the maximum
+	// buffer to place back in the pool. If the buffer is larger than the
+	// limit, we drop the buffer and not place back.
+	if buf.Cap() >= 64*1024 {
```

Comment:
This capacity limit threshold refers to golang's `fmt.Sprintf`, see https://github.com/golang/go/blob/80143607f06fd6410700e9764cfea9aaac9c311c/src/fmt/print.go#L169

Reply:
@hzxuzhonghu There is an explanation for why we need this. It is only useful when buffers of different capacities share the same pool.

Reply:
In worst case, there could be all buffer empty with 64 KB size there

Reply:
Would it be better to change it like this ? or remove it ?
```
if buf.Cap() >= 64*1024 && buf.Cap() >= p.cap * 2 {
  buf = nil
  return
}
```

Reply:
I think sync pool is better used for known fixed size case, but not for all cases. 

Here i am concerned about the effect

Reply:
I don't think so. Golang's `fmt.Sprintf` is a case of unfixed size. I have also seen other cases such as hash calculation and log string concatenation in other projects.

In fact, in the worst case: if the buffer is not put back into the pool, a buffer object will be created when calling `Get`, it should not be much different in performance than not using a pool(creating a buffer object each time) at all.

Reply:
hmm, i have seen many articles say it is not recommended to use fmt.Sprintf becuase of its performance

Reply:
Performance and using unfixed size seem to be different issues, can you give some suggestions?

Reply:
I think it seems reasonable not to put buffers that exceed the threshold back into the pool, which allows the GC to clean up the buffer quickly and avoid OOM.